lib/cn/kvset_split.c

```diff
@@ -0,0 +1,619 @@
+        struct key_stats stats = { 0 };
+        size_t off = 0, kmd_cnt_off;
+        uint nvals;
+
+        wbti_prefix(wbti, &cur.ko_pfx, &cur.ko_pfx_len);
+
+        if (HSE_UNLIKELY(start && key_obj_cmp(&cur, start) == 0))
+            continue;
+        start = NULL;
+
+        if (HSE_UNLIKELY(end && key_obj_cmp(&cur, end) > 0))
+            break;
+
+        nvals = stats.nvals = kmd_count(kmd, &off);
+        kmd_cnt_off = off;
+
+        while (nvals--) {
+            struct kvs_vtuple_ref vref = { 0 };
+            u64 vseq;
+
+            /* Pass NULL for vgmap as vbidx is not used here */
+            wbt_read_kmd_vref(kmd, NULL, &off, &vseq, &vref);
+
+            switch (vref.vr_type) {
+            case vtype_val:
+            case vtype_cval:
+                stats.tot_vlen += (vref.vb.vr_complen ? : vref.vb.vr_len);
+                break;
+
+            case vtype_ival:
+                stats.tot_vlen += vref.vi.vr_len;
+                break;
+
+            case vtype_tomb:
+                ++stats.ntombs;
+                break;
+
+            case vtype_zval:
+                break;
+
+            case vtype_ptomb:
+            default:
+                assert(0);
+                break;
+            }
+        }
+
+        err = kbb_add_entry(kbb, &cur, kmd + kmd_cnt_off, off - kmd_cnt_off, &stats);
+        if (!err)
+            ++tot_keys;
+    }
+
+    if (!err && tot_keys > 0) {
+        err = kbb_finish(kbb, kblks_out);
+        if(ev(!err && kblks_out->n_blks > 1)) {
+            log_info("output kblks %d > 1", kblks_out->n_blks);
+        }
+    }
+
+    wbti_destroy(wbti);
+    kbb_destroy(kbb);
+
+    return err;
+}
+
+/**
+ * kblock_split() - split a given kblock (@kbd) into two kblocks at @split_key
+ *
+ * @kbd:        kblock descriptor describing the source kblock
+ * @split_key:  the key at which the kblock needs to be split.
+ * @kblks_left:  (output) blk list of kblocks containing keys <= split key
+ * @kblks_right: (output) blk list of kblocks containing keys > split key
+ *
+ * NOTE:
+ * If either @kbid_left or @kbid_right is not populated and err == 0, then all
+ * keys from the source kblock got written to either @kbid_left or @kbid_right
+ *
+ * The output mblocks (@kbid_left and @kbid_right) are not committed when this
+ * function returns. It is up to the caller to abort or commit them.
+ */
+static merr_t
+kblock_split(
+    struct kblock_desc   *kbd,
+    const struct key_obj *split_key,
+    struct blk_list      *kblks_left,
+    struct blk_list      *kblks_right)
+{
+    merr_t err;
+
+    INVARIANT(kbd && split_key && kblks_left && kblks_right);
+
+    err = kblock_copy_range(kbd, NULL, split_key, kblks_left);
+    if (!err) {
+        err = kblock_copy_range(kbd, split_key, NULL, kblks_right);
+        if (!err && (kblks_left->n_blks == 0 && kblks_right->n_blks == 0)) {
+            assert(kblks_left->n_blks > 0 || kblks_right->n_blks > 0);
+            err = merr(EBUG);
```

Comment:
How serious is the problem?  abort() causes log replay on restart which can potentially be hazardous to data integrity, so best only to call abort() when thing's are really hosed...

Reply:
I just feel like we use EBUG in places where we should never be.

Reply:
there's a whole debate we can have here....  ideally we would handle the error, and in this case, fail the split operation, which may result in the KVDB limping along but at least it would be functional and not corrupt.

Reply:
I think EBUG makes sense here. I don't see any other reason why the left and right kblocks would be empty as a kblock cannot exist without having at least one key?